### PR TITLE
feat: add before send callback

### PR DESCRIPTION
**Pull request by marandaneto**
## :bulb: Motivation and Context

Customers need a configuration hook to inspect, modify, or drop events before upload, matching the before-send behavior available in other PostHog SDKs.

This adds a `:before_send` config callback. It receives the fully enriched event map after global/system properties, UUID, timestamp, and redaction have been applied. Returning `nil` drops the event.

## :green_heart: How did you test it?

- `mix format --check-formatted`
- `mix test test/posthog_test.exs`
- `mix posthog.public_api --check`

## :pencil: Checklist

- [x] I reviewed the submitted code.
- [x] I added tests to verify the changes.
- [ ] I updated the docs if needed.
- [x] No breaking change or entry added to the changelog.

### If releasing new changes

- [ ] Ran `sampo add` to generate a changeset file

## 🤖 Agent context

**Autonomy:** Human-driven (agent-assisted)

Implemented with the pi coding agent. The callback is applied just before handing the enriched event to the sender, with invalid return values or raised exceptions dropping only that event.

### CODE REVIEW

File: lib/posthog.ex [hunk 1]
```diff
@@ -78,7 +80,36 @@ defmodule PostHog do
       properties: properties
     }
 
-    PostHog.Sender.send(event, name)
+    case run_before_send(Map.get(config, :before_send), event) do
```
[the remark this answers is not in the document]
Comment (marandaneto) [reply lib/posthog.ex]: Good catch! Fixed in 3c5f4be — Switched back to config.before_send now that disabled_config includes the key.

File: test/posthog_test.exs [hunk 2]
```diff
@@ -86,6 +86,27 @@ defmodule PostHogTest do
...
+      assert properties[:saw_fully_enriched_event] == true
+      refute Map.has_key?(properties, :secret)
+    end
+
+    @tag config: [before_send: &__MODULE__.drop_before_send/1, supervisor_name: PostHog]
+    test "before_send can drop events" do
+      assert :ok = PostHog.bare_capture("case tested", "distinct_id")
+
+      assert [] = all_captured()
+    end
```
[the remark this answers is not in the document]
Comment (marandaneto) [reply test/posthog_test.exs:130]: Good catch! Fixed in 3c5f4be — Added generated tests for nil, invalid return, and raised callback paths.

**Comment by marandaneto** [comment]
Closes https://github.com/PostHog/posthog-elixir/issues/144

File: lib/posthog.ex [hunk 3]
```diff
@@ -78,7 +80,36 @@ defmodule PostHog do
...
+      nil ->
+        nil
+
+      %{} = event ->
+        event
+
+      other ->
+        Logger.error(
+          "PostHog before_send callback returned #{inspect(other)} instead of an event map or nil; dropping event"
+        )
```
Comment (martosaur) [inline lib/posthog.ex:105]: This feels problematic, because logging an error also creates an error tracking event. So if `before_send` is consistently failing, we'll have a infinite loop on our hands. Perhaps the fallback behavior should be to send the original event?

File: lib/posthog.ex [hunk 4]
```diff
@@ -78,7 +80,36 @@ defmodule PostHog do
...
+
+        nil
+    end
+  rescue
```
Comment (ioannisj) [inline lib/posthog.ex:109]: robot says we'll probably also need a catch clause here along with rescue throw/exit from hook

**Review by ioannisj — APPROVED** [review]
LG, couple of comments to review

File: lib/posthog.ex (hunk 3 above)
Comment (marandaneto) [reply lib/posthog.ex:105]: oh yeah good catch with `capture_level` enabled, not intentional, will fix it

Comment (marandaneto) [reply lib/posthog.ex:105]: Fixed in caa5cef by removing Logger calls from before_send failure handling. Explicit nil still drops; invalid return values or raised callbacks now send the original event to avoid recursive error capture.

File: lib/posthog.ex (hunk 4 above)
Comment (marandaneto) [reply lib/posthog.ex:109]: Yes, this was a real edge case: `rescue` only handles raised exceptions, not `throw`/`exit`. Fixed in fa874cf by adding a `catch` clause and test coverage for callbacks that throw or exit.
